lkmpg.tex

```diff
@@ -1204,6 +1204,66 @@ \section{System Calls}
 
+To change the pointer in \cpp|sys_call_table|, we need to consider the control register.
+A control register is a processor register that changes or controls the general behavior of the CPU.
+For x86 architecture, the cr0 register has various control flags that modify the basic operation of the processor.
+The WP flag in cr0 is write protection.
+Once the WP flag is set, the CPU may not write into the read-only sections.
+To modify \cpp|sys_call_table| we need to disable the WP flag.
+Since Linux v5.3, the \cpp|write_cr0| function cannot be used because of the sensitive cr0 bits pinned by the security issue, so we have to provide the custom asm to bypass it.
+However, \cpp|sys_call_table| symbol was unexported to prevent misuse.
+But there have few ways to get the symbol, memory seeking, and \cpp|kallsyms_lookup_name|, here we use \cpp|kallsyms_lookup_name|.
+Unfortunately, since Linux v5.11, \cpp|kallsyms_lookup_name| is also unexported, it needs certain trick to get the address of \cpp|kallsyms_lookup_name|.
+If \cpp|CONFIG_KPROBES| is enabled, we use Kprobes to insert the breakpoint to get the corresponding function address.
+We can facilitate the retrieval of function addresses by means of Kprobes to dynamically break into the specific kernel routine.
+Kprobes inserts a breakpoint at the entry of function by replacing the first bytes of the probed instruction.
+When a CPU hits the breakpoint, registers are stored, and the control will pass to Kprobes.
+It passes the addresses of the saved registers and the Kprobe struct to the handler you defined, then executes it.
+Kprobes can be registered by symbol name or address.
+Within the symbol name, the address will be handled by the kernel.
+Otherwise, use manual symbol lookup or getting the address of \cpp|sys_call_table| from \verb|/proc/kallsyms| and \verb|/boot/System.map|.
+Following are finding the address from \verb|/proc/kallsyms|:
+\begin{verbatim}
+$ sudo grep sys_call_table /proc/kallsyms
+ffffffff82000280 R x32_sys_call_table
+ffffffff820013a0 R sys_call_table
+ffffffff820023e0 R ia32_sys_call_table
+$ insmod syscall.ko addr=0xffffffff820013a0
+\end{verbatim}
+Using the address from \verb|/boot/System.map|, be careful about the KASLR (Kernel Address Space Layout Randomization).
+The KASLR may randomize the kernel code at every boot time.
+If \verb|nokaslr| doesn't set, the KASLR is enabled in your machine:
+\begin{verbatim}
+$ grep GRUB_CMDLINE_LINUX_DEFAULT /etc/default/grub
+GRUB_CMDLINE_LINUX_DEFAULT="quiet splash"
+$ sudo grep sys_call_table /proc/kallsyms
+ffffffff820013a0 R sys_call_table
+# Reboot
+$ sudo grep sys_call_table /proc/kallsyms
+ffffffff82000300 R sys_call_table
+\end{verbatim}
+So, under the KASLR, each time using the address from \verb|/proc/kallsyms| may have to take care of the address you used is from the boot time you found.
```

Comment:
Change into following:
If KASLR is enabled, we have to take care of address from \verb|/proc/kallsyms| each time we reboot the machine.

Reply:
Actually I don’t understand this paragraph very well.
In the line 1233, you mentioned about the defector of `/boot/System.map` if KALSR is enabled.
However the following example you provide (line1236~1244) is about `/proc/kallsyms`.
Do I misunderstand anything?

Reply:
I improve the explanation. The relationship between KASLR and /boot/System.map see [this](https://bneuburg.github.io/volatility/kaslr/2017/04/26/KASLR1.html).

Reply:
I see, thanks for the explanation.